modules/renter/contractor/allowance.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	// check that allowance is sufficient to store at least one sector
+	numSectors, err := maxSectors(a, c.hdb)
+	if err != nil {
+		return err
+	} else if numSectors == 0 {
+		return errInsufficientAllowance
+	}
+
+	c.mu.RLock()
+	shouldRenew := a.Period != c.allowance.Period || a.Funds.Cmp(c.allowance.Funds) != 0
+	remaining := int(a.Hosts) - len(c.contracts)
+	c.mu.RUnlock()
+
+	// if no contracts need renewing, run special-case logic instead
+	if !shouldRenew {
+		return c.managedFormAllowanceContracts(remaining, numSectors, a)
+	}
+
+	c.mu.RLock()
+	// gather contracts to renew
+	var renewSet []modules.RenterContract
+	for _, contract := range c.contracts {
+		renewSet = append(renewSet, contract)
+	}
+
+	// calculate new endHeight; if the period has not changed, the endHeight
+	// should not change either
+	endHeight := c.blockHeight + a.Period
+	if a.Period == c.allowance.Period {
+		for _, contract := range c.contracts {
+			endHeight = contract.EndHeight()
+			break
+		}
+	}
+	c.mu.RUnlock()
+
+	// renew existing contracts with new allowance parameters
+	newContracts := make(map[types.FileContractID]modules.RenterContract)
+	for _, contract := range renewSet {
+		newContract, err := c.managedRenew(contract, numSectors, endHeight)
+		if err != nil {
+			c.log.Printf("WARN: failed to renew contract with %v; a new contract will be formed in its place", contract.NetAddress)
+			remaining++
+			continue
+		}
+		newContracts[newContract.ID] = newContract
+		if len(newContracts) >= int(a.Hosts) {
+			break
+		}
+	}
+
+	// if we did not renew enough contracts, form new ones
+	if remaining > 0 {
+		formed, err := c.managedFormContracts(remaining, numSectors, endHeight)
+		if err != nil {
+			return err
+		}
+		for _, contract := range formed {
+			newContracts[contract.ID] = contract
+		}
+	}
+
+	// if we weren't able to form anything, return an error
+	if len(newContracts) == 0 {
+		return errors.New("unable to form or renew any contracts")
+	}
+
+	// Set the allowance and replace the contract set
+	c.mu.Lock()
+	c.allowance = a
+	c.contracts = newContracts
+	// update metrics
+	var spending types.Currency
+	for _, contract := range c.contracts {
+		spending = spending.Add(contract.RenterFunds())
+	}
+	c.financialMetrics.ContractSpending = spending
+	err = c.saveSync()
+	c.mu.Unlock()
+
+	return err
+}
+
+// managedFormAllowanceContracts handles the special case where no contracts
+// need to be renewed when setting the allowance.
+func (c *Contractor) managedFormAllowanceContracts(n int, numSectors uint64, a modules.Allowance) error {
+	if n <= 0 {
+		return nil
+	}
+
+	// if we're forming contracts but not renewing, the new contracts should
+	// have the same endHeight as the existing ones. Otherwise, the endHeight
+	// should be a full period.
+	c.mu.RLock()
+	endHeight := c.blockHeight + a.Period
+	for _, contract := range c.contracts {
+		endHeight = contract.EndHeight()
+		break
+	}
```

Comment:
Might be good to throw in a sanity check then. If you are doing atomic swaps it shouldn't be a problem, but it's better to have code which reassures you it's not a problem then to have stuff silently failing, especially since you might forget about this 4 months down the line when some change we make starts causing problems.